source/adapters/level_zero/v2/event_pool.hpp

```diff
@@ -0,0 +1,49 @@
+//===--------- event_pool.hpp - Level Zero Adapter ------------------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <stack>
+
+#include <unordered_map>
+#include <ur/ur.hpp>
+#include <ur_api.h>
+#include <vector>
+#include <ze_api.h>
+
+#include "../device.hpp"
+#include "common.hpp"
+#include "event.hpp"
+#include "event_provider.hpp"
+
+namespace v2 {
+
+class event_pool {
+public:
+  event_pool(std::unique_ptr<event_provider> Provider)
+      : Provider(std::move(Provider)){};
+  event_pool(event_pool &&other) : Provider(std::move(other.Provider)) {}
+
+  ~event_pool(){};
+
+  DeviceId Id() { return Provider->device()->Id; };
+
+  ur_event *allocate();
```

Comment:
Shouldn;'t we make this method thread safe?

Reply:
No, because `event_pool` is meant to be stored in the queue, under its lock. And the queue **must** outlive all the events, so theoretically there should be no problem in waiting with event deallocation until the queue is finished.